### PR TITLE
451006 aliaksandr kozlov

### PR DESCRIPTION
**Pull request by mrblamber**

### CODE REVIEW

File: Tracer/TraceResult.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
+﻿using System.Collections.Generic;
+using System.Diagnostics;
+using System.Reflection;
+using System.Threading;
+
+namespace Tracer
+{
+    public class TraceResult
+    {
+        private static readonly object LockMethodAdd = new object();
```
Comment (RamanBut-Husaim) [inline Tracer/TraceResult.cs]: @mrblamber Could you please explain why are you using separate lock objects for add and remove? The usage of separate locks could lead to modifications to your internal collections by two threads at a time. In add your `_currentNodes` will be modified by one thread and at the same time the second thread could try to read from this collection.

File: Tracer/Tracer.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+{
+    public class Tracer : ITracer
+    {
+        public TraceResult TraceResult { get; }
+
+        private static Tracer _instance;
+
+        private static readonly object LockObject = new object();
+
+        private static readonly object StartTraceObject = new object();
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer.cs]: @mrblamber The same lock approach here.

File: Tracer/XmlFormatter.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+                var thread = new XElement("thread", new XAttribute("id", element.Key));
+                foreach (var node in element.Value)
+                {
+                    AddMethodToXmlTree(thread, node);
+                }
+                root.Add(thread);
+            }
+
+            document.Add(root);
+            document.Save("D:\\batka.xml");
```
Comment (RamanBut-Husaim) [inline Tracer/XmlFormatter.cs]: @mrblamber Accept a file path as a constructor parameter.

File: Tracer/Tracer.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+        private static readonly object StopTraceObject = new object();
+
+        public static Tracer Instance
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer.cs]: @mrblamber Read about other [possible singleton implementations](http://csharpindepth.com/Articles/General/Singleton.aspx).

File: Tracer/Tracer.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+                        {
+                            _instance = new Tracer();
+                        }
+                    }
+                }
+                return _instance;
+            }
+        }
+
+        protected Tracer()
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer.cs]: @mrblamber Why the constructor is `protected`?

File: Tracer/TraceResult.cs (hunk 1 above)
Comment (mrblamber) [inline Tracer/TraceResult.cs]: @RamanBut-Husaim Yes, my bad. I'll fix it.

File: Tracer/Tracer.cs (hunk 4 above)
Comment (mrblamber) [inline Tracer/Tracer.cs]: @RamanBut-Husaim Have I fix it?

Comment (RamanBut-Husaim) [inline Tracer/Tracer.cs]: @mrblamber Nope. Just take a look. I could ask you about them. :)

File: Tracer/Tracer.cs (hunk 5 above)
Comment (mrblamber) [inline Tracer/Tracer.cs]: @RamanBut-Husaim It's because initially I wanted to extend this class by another. Bad Idea.

**Comment by RamanBut-Husaim** [comment]
@mrblamber Please, leave a comment when you finish all your fixes.

File: Tracer/XmlFormatter.cs (hunk 3 above)
Comment (mrblamber) [inline Tracer/XmlFormatter.cs]: @RamanBut-Husaim Please, take a look at my fixes.

**Comment by mrblamber** [comment]
@RamanBut-Husaim Please, take a look at my fixes.

File: Tracer/TraceResult.cs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+        public TraceResult()
+        {
+            TraceTree = new Dictionary<int, List<MethodsTreeNode>>();
+            _currentNodes = new Dictionary<int, MethodsTreeNode>();
+        }
+
+        public void AddMethodToTree(MethodBase methodBase)
+        {
+            lock (LockObject)
```
Comment (RamanBut-Husaim) [inline Tracer/TraceResult.cs]: @mrblamber Are you sure that you need a lock there? It seems that all operations in this class are performed in the scope of lock from `Tracer` class.

Comment (mrblamber) [inline Tracer/TraceResult.cs]: @RamanBut-Husaim Yes, again my bad.

**Comment by mrblamber** [comment]
@RamanBut-Husaim Please, take a look at my fixes.

File: Tracer/packages.config [hunk 7]
```diff
@@ -0,0 +1,4 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="Newtonsoft.Json" version="9.0.1" targetFramework="net452" />
```
Comment (RamanBut-Husaim) [inline Tracer/packages.config]: @mrblamber Why are you using this?

Comment (mrblamber) [inline Tracer/packages.config]: @RamanBut-Husaim Because initially I wanted to use Json Formatter.

**Comment by mrblamber** [comment]
@RamanBut-Husaim Thank you!

**Comment by mrblamber** [comment]
@RamanBut-Husaim All changes are done.

File: Tracer/Tracer.csproj [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+    <DebugType>pdbonly</DebugType>
+    <Optimize>true</Optimize>
+    <OutputPath>bin\Release\</OutputPath>
+    <DefineConstants>TRACE</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
+  </PropertyGroup>
+  <ItemGroup>
+    <Reference Include="Newtonsoft.Json, Version=9.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed, processorArchitecture=MSIL">
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer.csproj]: @mrblamber Why are you referencing Newtonsoft.Json?

Comment (mrblamber) [inline Tracer/Tracer.csproj]: @RamanBut-Husaim Wow, I've deleted it from references. It's because  I wanted to use Json Formatter.

**Comment by RamanBut-Husaim** [comment]
@mrblamber Approved.

**Comment by mrblamber** [comment]
@RamanBut-Husaim Finished developing.

**Comment by RamanBut-Husaim** [comment]
@mrblamber Accepted.

**Pull request merged**